File: python/01_basics/variables.py
"""
Python Basics - Variables and Data Types
"""

# Basic Types
number_variable = 42
string_variable = "Hello, Python!"
boolean_variable = True
none_variable = None

# Lists
number_list = [1, 2, 3, 4, 5]
string_list = ["a", "b", "c"]

# Tuples
tuple_example = ("age", 30)

# Dictionaries
person = {
    "name": "John",
    "age": 30,
    "is_active": True
}

# Set
unique_numbers = {1, 2, 3, 4, 5, 5}  # Duplicates are automatically removed

# Type Hints (Python 3.5+)
from typing import List, Dict, Union, Tuple, Optional

# Function with type hints
def add(a: int, b: int) -> int:
    return a + b

# Complex type hints
def process_data(items: List[str], config: Dict[str, Union[str, int]]) -> Tuple[int, Optional[str]]:
    # This is just a demonstration function
    return (len(items), config.get("message"))

# Usage examples
if __name__ == "__main__":
    print(f"Number: {number_variable}")
    print(f"String: {string_variable}")
    print(f"Person: {person['name']}, {person['age']} years old")
    print(f"Unique numbers: {unique_numbers}")
    print(f"2 + 3 = {add(2, 3)}")
    
    # Example with type hints
    result, message = process_data(
        ["item1", "item2"],
        {"message": "Success", "count": 2}
    )
    print(f"Processed {result} items. Message: {message}") 

File: typescript/01_basics/variables.ts
/**
 * TypeScript Basics - Variables and Data Types
 */

// Basic Types
const numberVariable: number = 42;
const stringVariable: string = "Hello, TypeScript!";
const booleanVariable: boolean = true;
const undefinedVariable: undefined = undefined;
const nullVariable: null = null;

// Arrays
const numberArray: number[] = [1, 2, 3, 4, 5];
const stringArray: Array<string> = ["a", "b", "c"];

// Tuples
const tuple: [string, number] = ["age", 30];

// Object Types
interface Person {
  name: string;
  age: number;
  isActive?: boolean; // Optional property
}

const person: Person = {
  name: "John",
  age: 30,
  isActive: true
};

// Union Types
let id: string | number = 123;
id = "ABC-123"; // Also valid

// Type Assertions
const someValue: any = "This is a string";
const strLength: number = (someValue as string).length;

// Enums
enum Direction {
  Up = "UP",
  Down = "DOWN",
  Left = "LEFT",
  Right = "RIGHT"
}
const playerDirection: Direction = Direction.Up;

// Function with type annotations
function add(a: number, b: number): number {
  return a + b;
}

// Usage examples
console.log(`Number: ${numberVariable}`);
console.log(`String: ${stringVariable}`);
console.log(`Person: ${person.name}, ${person.age} years old`);
console.log(`Direction: ${playerDirection}`);
console.log(`2 + 3 = ${add(2, 3)}`); 